### PR TITLE
port to 2.0

### PR DESCRIPTION
**Pull request by notwaldorf**
Diffs sans whitespace: https://github.com/PolymerElements/paper-input/pull/445/files?w=1

- :warning: Changes :warning: Distribution is now done with slots, so things have changed because of that (before: `<label>foo</label>`, now `<label slot="label">foo</label>`, etc)
- all the `<input is="iron-input">` have been replaced by `<iron-input><input></iron-input`
- `preventInvalidInput` has merged with `allowedPattern`, similarly to how that was done in `iron-input`
- moved the auto-floating of the label if the native input has a placeholder to `paper-input` itself, rather that `PaperInputBehavior`, since it depends on the timing of the `IronInput` being attached
- replaced all the `forceXIfStamp(input);` in tests with a `Polymer.Base.async` call to wait for things to stamp

### CODE REVIEW

**Comment by mkazlauskas** [comment]
Switching focus with `tab` doesn't seem to be working. Any ideas on that?

**Comment by notwaldorf** [comment]
@mkazlauskas This is a known shadow DOM bug: crbug.com/618587

**Comment by mkazlauskas** [comment]
Do I need to do something regarding CLA? I have it signed, so not sure what is this about.

**Comment by notwaldorf** [comment]
hush that @googlebot, who else do you think authored those commits?

**Comment by 0xR** [comment]
The version of paper-input is "2.0.0" in this branch. This would indicate that it is a breaking change. However, it does look like it a hybrid component.

So is this a breaking change? If so, am I right to assume that the documentation on how to upgrade will be added in the future?

**Comment by notwaldorf** [comment]
@0xR All hybrid elements will be tagged as 2.0 as they will most likely contain very minimal breaking changes. For example, `paper-input#2.0.0` uses `slot` instead of `content`, so how you distribute a prefix (for example) is different. I don't think we would have a "how to upgrade" documentation for these elements since the changes are minimal and mostly covered by https://www.polymer-project.org/2.0/docs/devguide/hybrid-elements

**Comment by stramel** [comment]
@notwaldorf Looks like `iron-component-page` is missing in the `bower.json` but is still present in the `index.html`

**Comment by notwaldorf** [comment]
@e111077 when you get to reviewing this: there's something dodgy going around with the shady dom 2.0 focus tests -- the `focused` attribute from `iron-behaviors` isn't being set correctly :/

File: paper-input.html [hunk 1]
```diff
@@ -82,34 +81,80 @@
...
+      </template>
+
+    </paper-input-container>
+  </template>
 
-      <input is="iron-input" id="input"
+  <!-- This is a fresh new hell to make this element hybrid. Basically, in 2.0
+    we lost is=, so the example same template can't be used with iron-input 1.0 and 2.0.
+    Expect some conditional code (especially in the tests).
+   -->
```
Comment (e111077) [inline paper-input.html]: uhhhhhhh....

File: paper-input.html [hunk 2]
```diff
@@ -143,28 +188,102 @@
...
-      <template is="dom-if" if="[[errorMessage]]">
-        <paper-input-error aria-live="assertive">[[errorMessage]]</paper-input-error>
-      </template>
-
-      <template is="dom-if" if="[[charCounter]]">
-        <paper-input-char-counter></paper-input-char-counter>
-      </template>
-
-    </paper-input-container>
+  <template id="v1">
```
Comment (e111077) [inline paper-input.html]: 😢

File: paper-input.html [hunk 3]
```diff
@@ -143,28 +188,102 @@
...
 
     behaviors: [
-      Polymer.IronFormElementBehavior,
-      Polymer.PaperInputBehavior
-    ]
+      Polymer.PaperInputBehavior,
+      Polymer.IronFormElementBehavior
+    ],
+
+    // TODO: remove this when 2.0 adds beforeRegister back to legacy
```
Comment (e111077) [inline paper-input.html]: You should do the do

**Comment by e111077** [comment]
rebase plsss

File: demo/ssn-input.html [hunk 4]
```diff
@@ -34,63 +33,107 @@
...
+      beforeRegister: function() {
+        this._beforeRegister();
+      },
 
-    _computeValue: function(ssn1, ssn2, ssn3) {
-      this.value = ssn1.trim() + '-' + ssn2.trim() + '-' + ssn3.trim();
-    }
-  });
-</script>
+      _beforeRegister: function() {
```
Comment (notwaldorf) [inline demo/ssn-input.html]: Sorry, Kevin had briefly removed `beforeRegister` from 2.0, but I think it's back now. Will fix.

File: paper-input-behavior.html [hunk 5]
```diff
@@ -85,15 +85,6 @@
       },
 
       /**
-       * Set to true to prevent the user from entering invalid input. If you're
-       * using PaperInputBehavior to  implement your own paper-input-like element,
-       * bind this to `<input is="iron-input">`'s `preventInvalidInput` property.
-       */
-      preventInvalidInput: {
```
Comment (notwaldorf) [inline paper-input-behavior.html]: They were both either used together, or not at all. (like, you can't just use one without the other). Did I misunderstand it?

Comment (e111077) [inline paper-input-behavior.html]: Yeah, I agree that it's insignificant.

File: demo/ssn-input.html (hunk 4 above)
Comment (notwaldorf) [inline demo/ssn-input.html]: Ugh @sorvell says it's not back yet 😭

Comment (e111077) [inline demo/ssn-input.html]: sosad please `TODO`

File: .travis.yml [hunk 6]
```diff
@@ -1,24 +1,28 @@
...
     sources:
       - google-chrome
     packages:
-      - google-chrome-stable
+      - google-chrome-beta
+  sauce_connect: true
 script:
-  - xvfb-run wct
-  - 'if [ "${TRAVIS_PULL_REQUEST}" = "false" ]; then wct -s ''default''; fi'
+  - xvfb-run polymer test
```
Comment (stramel) [inline .travis.yml]: Does this test both variants?

Comment (notwaldorf) [inline .travis.yml]: yup

File: paper-input.html [hunk 7]
```diff
@@ -82,6 +81,27 @@
...
+        /* Firefox sets a min-width on the input, which can cause layout issues */
+        min-width: 0;
+
+        @apply --paper-font-subhead;
+        @apply --paper-input-container-input;
+      }
+      
       input::-webkit-outer-spin-button,                                                                                                     
       input::-webkit-inner-spin-button {                                                                                                    
         @apply(--paper-input-container-input-webkit-spinner);
```
Comment (bicknellr) [inline paper-input.html]: a few stray cases of `@apply(x);` -> `@apply x;` here

File: paper-input-container.html [hunk 8]
```diff
@@ -240,97 +263,85 @@
...
+      .input-content ::slotted(.paper-input-input) {
+        @apply --paper-input-container-shared-input-style;
+        @apply --paper-input-container-input;
       }
 
-      .input-content ::content input::-webkit-outer-spin-button,
-      .input-content ::content input::-webkit-inner-spin-button {
+      .input-content ::slotted(input)::-webkit-outer-spin-button,
+      .input-content ::slotted(input)::-webkit-inner-spin-button {
         @apply(--paper-input-container-input-webkit-spinner);
```
Comment (bicknellr) [inline paper-input-container.html]: Here's another `@apply` w/ parens.

File: demo/ssn-input.html (hunk 4 above)
Comment (notwaldorf) [inline demo/ssn-input.html]: todo already exists! it's on the `registered` method

File: paper-input.html (hunk 7 above)
Comment (notwaldorf) [inline paper-input.html]: thanks! stupid rebases 😳

File: paper-input.html (hunk 3 above)
Comment (notwaldorf) [inline paper-input.html]: Polymer didn't do the do

File: paper-input-container.html [hunk 9]
```diff
@@ -14,19 +14,25 @@
...
 For example:
 
     <paper-input-container>
-      <label>Your name</label>
-      <input is="iron-input">
+      <label slot="label">Your name</label>
+      <iron-input slot="input">
+        <input>
+      </iron-input>
+      // In Polymer 1.0, you would use `<input is="iron-input" slot="input">` instead ov the above.
```
Comment (notwaldorf) [inline paper-input-container.html]: Oh I think it messes up highlighting in Atom really badly :(

Comment (bicknellr) [inline paper-input-container.html]: also ` instead ov the above`

File: demo/ssn-input.html (hunk 4 above)
Comment (e111077) [inline demo/ssn-input.html]: you're so good that you can travel time

File: paper-input.html (hunk 3 above)
Comment (e111077) [inline paper-input.html]: sosad

Comment (notwaldorf) [inline paper-input.html]: !!! Polymer did the do since then so i can do the do!

**Comment by notwaldorf** [comment]
I can't get those 2 tests to pass on travis (something is stealing focus away from the tests), but they're hella green locally:
screen shot 2017-04-13 at 5 06 48

File: paper-textarea.html [hunk 10]
```diff
@@ -10,7 +10,6 @@
 
 <link rel="import" href="../polymer/polymer.html">
 <link rel="import" href="../iron-autogrow-textarea/iron-autogrow-textarea.html">
-<link rel="import" href="../iron-form-element-behavior/iron-form-element-behavior.html">
```
Comment (e111077) [inline paper-textarea.html]: why is this not a form-element now?

**Comment by e111077** [comment]
After this gets merged, we should figure out a more-permanent / -trackable place to put these breaking changes and api changes

File: paper-textarea.html (hunk 10 above)
Comment (notwaldorf) [inline paper-textarea.html]: because i did a dumb

**Comment by notwaldorf** [comment]
@e111077 done! also, yes, re: tracking changes. however, for `paper-input` proper, there's basically no breaking changes...i think.

**Comment by e111077** [comment]
Color me impressed!

**Comment by bicknellr** [comment]
`@apply(...);`:
```
./paper-input-container.html:307:        @apply(--paper-input-container-input-focus);  
./paper-input-container.html:314:        @apply(--paper-input-container-input-invalid);
./paper-input.html:127:        @apply(--paper-input-container-ms-clear);               
```

**Pull request merged**